File: app/src/main/java/net/gerosyab/magicball/activity/MainActivity.java
/*
 *     TheClassicMagicBall - Android Magic 8 Ball Simulator
 *     Copyright (C) 2014 gerosyab
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.gerosyab.magicball.activity;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.os.Vibrator;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;

import net.gerosyab.magicball.R;
import net.gerosyab.magicball.data.Const;
import net.gerosyab.magicball.fragment.MainFragment;
import net.gerosyab.magicball.fragment.MsgFragment;
import net.gerosyab.magicball.util.MyLog;
import net.gerosyab.magicball.util.Shaker;

public class MainActivity extends AppCompatActivity implements Shaker.Callback {

    Context context;
    LinearLayout motherLinear;
    LinearLayout contentLinear;
    static Shaker shaker;

    MainFragment mainFragment;
    MsgFragment msgFragment;

    public static Vibrator vibrator;

    private AdView mAdView;

    private boolean mBackKeyFlag = false;

    @SuppressLint("HandlerLeak")
    private Handler mHandler = new Handler(){
        public void handleMessage(Message msg){
            if(msg.what == 0){
                mBackKeyFlag = false;
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        MobileAds.initialize(this, "ca-app-pub-4746845653236332~3245362320");

        mAdView = (AdView) findViewById(R.id.adView);
        mAdView.loadAd(new AdRequest.Builder().addTestDevice(AdRequest.DEVICE_ID_EMULATOR).build());

        context = getApplicationContext();

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);

        motherLinear = (LinearLayout) findViewById(R.id.mother_linear);
        contentLinear = (LinearLayout) findViewById(R.id.content_linear);

        mainFragment = new MainFragment();
        msgFragment = new MsgFragment();
        getSupportFragmentManager().beginTransaction().add(R.id.content_linear, mainFragment).commit();

    }

    @Override
    public void onResume() {
        MyLog.d("MainActivity", "onResume");
        super.onResume();

        if (mAdView != null) {
            mAdView.resume();
        }

        if(shaker == null){
            shaker = new Shaker(context, this);
        }
        shaker.open();
    }

    @Override
    public void onPause() {
        MyLog.d("MainActivity", "onPause");
        super.onPause();

        if (mAdView != null) {
            mAdView.pause();
        }

        if(shaker != null){
            shaker.close();
        }
    }

    /** Called before the activity is destroyed. */
    @Override
    public void onDestroy() {
        MyLog.d("MainActivity", "onDestroy");
        // Destroy the AdView.
        if (mAdView != null) {
            mAdView.destroy();
        }
        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        MyLog.d("MainActivity", "onBackPressed getSupportFragmentManager().getBackStackEntryCount() " + getSupportFragmentManager().getBackStackEntryCount());
        // Catch back action and pops from backstack
        // (if you called previously to addToBackStack() in your transaction)
        if (getSupportFragmentManager().getBackStackEntryCount() > 0){
            MyLog.d("MainActivity", "onBackPressed getSupportFragmentManager().getBackStackEntryCount() > 0");
            getSupportFragmentManager().popBackStack();
        }
        // Default action on back pressed
        else if (!mBackKeyFlag) {
            MyLog.d("MainActivity", "onBackPressed !mBackKeyFlag");
            Toast.makeText(context, "Back Again to Exit", Toast.LENGTH_SHORT).show();
            mBackKeyFlag = true;
            mHandler.sendEmptyMessageDelayed(0, 2000);
        }
        else {
            super.onBackPressed();
        }

    }

    @Override
    public void onShakingDetected() {
        MyLog.d("MainActivity", "onShakingDetected getSupportFragmentManager().getBackStackEntryCount() " + getSupportFragmentManager().getBackStackEntryCount());
        vibrator.vibrate(Const.vibTime);
        if (getSupportFragmentManager().getBackStackEntryCount() > 0){
            MyLog.d("MainActivity", "fragmentManager.getBackStackEntryCount() > 0");
            if(msgFragment != null) {
                MyLog.d("MainActivity", "setNewMessage()");
                msgFragment.setNewMessage();
            }
        }
        else {
            MyLog.d("MainActivity", "fragmentManager.getBackStackEntryCount() == 0");
//            transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE|FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
//            transaction.setTransitionStyle(FragmentTransaction.TRANSIT_FRAGMENT_FADE|FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
            getSupportFragmentManager().beginTransaction().replace(R.id.content_linear, msgFragment).addToBackStack(null).commitAllowingStateLoss();
            MyLog.d("MainActivity", "after transaction fragmentManager.getBackStackEntryCount() " + getSupportFragmentManager().getBackStackEntryCount());
        }
    }

    public static Shaker getShakerInstance(){
        return shaker;
    }
}

File: app/src/main/java/net/gerosyab/magicball/ui/MainActivity.kt
/*
 * TheClassicMagicBall - Android Magic 8 Ball Simulator
 * Copyright (C) 2014 DAISSUE
 */
package net.gerosyab.magicball.ui

import android.content.pm.ActivityInfo
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.commit
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.MobileAds
import net.gerosyab.magicball.BuildConfig
import net.gerosyab.magicball.R
import net.gerosyab.magicball.data.Const
import net.gerosyab.magicball.databinding.ActivityMainBinding
import net.gerosyab.magicball.util.MyLog
import net.gerosyab.magicball.util.Shaker

class MainActivity :
    AppCompatActivity(),
    Shaker.Callback {
    private lateinit var binding: ActivityMainBinding
    private var mBackKeyFlag = false
    private val handler = Handler(Looper.getMainLooper())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (!resources.getBoolean(R.bool.allow_rotation)) {
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        WindowCompat.setDecorFitsSystemWindows(window, false)
        ViewCompat.setOnApplyWindowInsetsListener(binding.motherLinear) { v, windowInsets ->
            val bars = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(bars.left, bars.top, bars.right, bars.bottom)
            WindowInsetsCompat.CONSUMED
        }

        if (BuildConfig.ADS_ENABLED) {
            MobileAds.initialize(this) {}
            binding.adView.loadAd(AdRequest.Builder().build())
        } else {
            binding.adBannerSlot.visibility = View.GONE
        }

        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        if (savedInstanceState == null) {
            supportFragmentManager.commit {
                replace(R.id.content_frame, MainFragment())
            }
        }
    }

    override fun onResume() {
        MyLog.d("MainActivity", "onResume")
        super.onResume()
        if (BuildConfig.ADS_ENABLED) binding.adView.resume()
        if (shaker == null) {
            shaker = Shaker(applicationContext, this)
        }
        shaker?.open()
    }

    override fun onPause() {
        MyLog.d("MainActivity", "onPause")
        if (BuildConfig.ADS_ENABLED) binding.adView.pause()
        shaker?.close()
        super.onPause()
    }

    override fun onDestroy() {
        MyLog.d("MainActivity", "onDestroy")
        if (BuildConfig.ADS_ENABLED) binding.adView.destroy()
        super.onDestroy()
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        MyLog.d(
            "MainActivity",
            "onBackPressed getSupportFragmentManager().getBackStackEntryCount() " +
                supportFragmentManager.backStackEntryCount,
        )
        if (supportFragmentManager.backStackEntryCount > 0) {
            MyLog.d("MainActivity", "onBackPressed popBackStack")
            supportFragmentManager.popBackStack()
            return
        }
        if (!mBackKeyFlag) {
            MyLog.d("MainActivity", "onBackPressed !mBackKeyFlag")
            Toast
                .makeText(
                    applicationContext,
                    R.string.exit_press_again,
                    Toast.LENGTH_SHORT,
                ).show()
            mBackKeyFlag = true
            handler.postDelayed({ mBackKeyFlag = false }, 2000)
        } else {
            super.onBackPressed()
        }
    }

    override fun onShakingDetected() {
        MyLog.d(
            "MainActivity",
            "onShakingDetected backStack=" + supportFragmentManager.backStackEntryCount,
        )
        if (supportFragmentManager.backStackEntryCount > 0) {
            // MsgFragment.setNewMessage() handles vibration itself
            (supportFragmentManager.findFragmentById(R.id.content_frame) as? MsgFragment)
                ?.setNewMessage()
        } else {
            vibrateShort()
            supportFragmentManager.commit {
                replace(R.id.content_frame, MsgFragment.newInstance(tabletScaleFactor = 1f))
                addToBackStack(null)
            }
        }
    }

    fun onBallTouched() {
        onShakingDetected()
    }

    fun openMsgFromPill() {
        vibrateShort()
        if (supportFragmentManager.backStackEntryCount > 0) {
            return
        }
        supportFragmentManager.commit {
            replace(R.id.content_frame, MsgFragment.newInstance(tabletScaleFactor = 1f))
            addToBackStack(null)
        }
    }

    fun vibrateShort() {
        val duration = Const.VIB_TIME_MS
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vm = getSystemService(VIBRATOR_MANAGER_SERVICE) as VibratorManager
            val v = vm.defaultVibrator
            v.vibrate(
                VibrationEffect.createOneShot(
                    duration,
                    VibrationEffect.DEFAULT_AMPLITUDE,
                ),
            )
        } else {
            @Suppress("DEPRECATION")
            val v = getSystemService(VIBRATOR_SERVICE) as Vibrator
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                v.vibrate(
                    VibrationEffect.createOneShot(
                        duration,
                        VibrationEffect.DEFAULT_AMPLITUDE,
                    ),
                )
            } else {
                @Suppress("DEPRECATION")
                v.vibrate(duration)
            }
        }
    }

    companion object {
        @Volatile
        @JvmField
        var shaker: Shaker? = null

        @JvmStatic
        fun getShakerInstance(): Shaker? = shaker
    }
}
